File: lang/lsp-server/src/main/java/org/xvm/lsp/model/Location.java
package org.xvm.lsp.model;

import org.jspecify.annotations.NonNull;

/**
 * Immutable source location.
 */
public record Location(
        @NonNull String uri,
        int startLine,
        int startColumn,
        int endLine,
        int endColumn
) {
    public Location {
        if (startLine < 0 || startColumn < 0 || endLine < 0 || endColumn < 0) {
            throw new IllegalArgumentException("Positions must be non-negative");
        }
    }

    public static Location of(
            final @NonNull String uri,
            final int line,
            final int column) {
        return new Location(uri, line, column, line, column);
    }

    public static Location ofLine(final @NonNull String uri, final int line) {
        return new Location(uri, line, 0, line, Integer.MAX_VALUE);
    }
}

File: lang/lsp-server/src/main/kotlin/org/xvm/lsp/model/Location.kt
package org.xvm.lsp.model

/**
 * Immutable source location.
 */
data class Location(
    val uri: String,
    val startLine: Int,
    val startColumn: Int,
    val endLine: Int,
    val endColumn: Int
) {
    init {
        require(startLine >= 0) { "startLine must be non-negative" }
        require(startColumn >= 0) { "startColumn must be non-negative" }
        require(endLine >= 0) { "endLine must be non-negative" }
        require(endColumn >= 0) { "endColumn must be non-negative" }
    }

    companion object {
        fun of(uri: String, line: Int, column: Int): Location =
            Location(uri, line, column, line, column)

        fun ofLine(uri: String, line: Int): Location =
            Location(uri, line, 0, line, Int.MAX_VALUE)
    }
}
